src/control/pbin/exec_test.go

```diff
@@ -38,35 +40,78 @@ import (
 
 func reqRes() {
 	conn := pbin.NewStdioConn("child", "parent", os.Stdin, os.Stdout)
-	readBuf := make([]byte, pbin.MaxMessageSize)
 
-	readLen, err := conn.Read(readBuf)
+	readBuf, err := pbin.ReadMessage(conn)
 	if err != nil {
 		childErrExit(err)
 	}
 
 	var req pbin.Request
-	err = json.Unmarshal(readBuf[:readLen], &req)
-	if err != nil {
+	if err = json.Unmarshal(readBuf, &req); err != nil {
 		childErrExit(err)
 	}
 
-	res := pbin.Response{
-		Payload: req.Payload,
+	var writeBuf []byte
+	switch req.Method {
+	case "garbage":
+		writeBuf = []byte(`junk`)
+	case "oversize":
+		writeBuf = generatePayload(pbin.MaxMessageSize + 1)
+	default:
+		res := pbin.Response{
+			Payload: req.Payload,
+		}
+
+		writeBuf, err = json.Marshal(&res)
+		if err != nil {
+			childErrExit(err)
+		}
 	}
 
-	writeBuf, err := json.Marshal(&res)
-	if err != nil {
+	if _, err = conn.Write(writeBuf); err != nil {
 		childErrExit(err)
 	}
 
-	_, err = conn.Write(writeBuf)
-	if err != nil {
+	if err = conn.CloseWrite(); err != nil {
 		childErrExit(err)
 	}
 }
 
-func TestPbinExec(t *testing.T) {
+func generatePayload(size int) []byte {
+	var buf bytes.Buffer
+	buf.WriteString(`{"reply":"`)
+	post := `"}`
+	adjSize := size - len(post)
+
+	for buf.Len() < adjSize {
+		buf.WriteString("0123456789abcdefghijklmnopqrstuvxyz")
+	}
+
+	ret := buf.Bytes()
+	if len(ret) > adjSize {
+		ret = ret[:len(ret)-(len(ret)-(adjSize))]
```

Comment:
Doesn't that simplify down to adjSize?

Reply:
... yes. Not really sure what I was thinking there. Fixed, thanks.